editoast/src/models/errors/routes.rs

```diff
@@ -1,12 +1,93 @@
+use std::fmt;
+
 use diesel::sql_types::{Array, Integer, Json, Text};
 use diesel::{sql_query, PgConnection, RunQueryDsl};
+use serde::{Deserialize, Serialize};
+use strum::IntoEnumIterator;
+use strum_macros::EnumIter;
 
 use super::InfraError;
-use crate::railjson::ObjectType;
+use crate::railjson::{Direction, DirectionalTrackRange, ObjectType, Route};
 use crate::{infra_cache::InfraCache, railjson::ObjectRef};
 use diesel::result::Error as DieselError;
 use serde_json::to_value;
 
+#[derive(Serialize, Deserialize, Debug, Clone, Copy, EnumIter)]
+#[serde(deny_unknown_fields)]
+pub enum PathEndpointField {
+    #[serde(rename = "entry_point")]
+    EntryPoint,
+    #[serde(rename = "exit_point")]
+    ExitPoint,
+}
+
+impl PathEndpointField {
+    /// Given a path, retrieve track id and position offset of the path endpoint location
+    pub fn get_path_location(&self, path: &[DirectionalTrackRange]) -> (String, f64) {
+        let track_range = match self {
+            PathEndpointField::EntryPoint => path.first().unwrap(),
+            PathEndpointField::ExitPoint => path.last().unwrap(),
+        };
+
+        let pos = match (self, &track_range.direction) {
+            (PathEndpointField::EntryPoint, Direction::StartToStop) => track_range.begin,
+            (PathEndpointField::EntryPoint, Direction::StopToStart) => track_range.end,
+            (PathEndpointField::ExitPoint, Direction::StartToStop) => track_range.end,
+            (PathEndpointField::ExitPoint, Direction::StopToStart) => track_range.begin,
+        };
+
+        (track_range.track.obj_id.clone(), pos)
+    }
+
+    pub fn get_route_endpoint<'a>(&self, route: &'a Route) -> &'a ObjectRef {
+        match self {
+            PathEndpointField::EntryPoint => &route.entry_point,
+            PathEndpointField::ExitPoint => &route.exit_point,
+        }
+    }
+}
+
+impl fmt::Display for PathEndpointField {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", serde_json::to_string(self).unwrap())
+    }
+}
+
+fn get_object(object: &ObjectRef, infra_cache: &InfraCache) -> Option<(String, f64)> {
+    let obj_id = &object.obj_id;
+    match object.obj_type {
+        ObjectType::Detector => infra_cache
+            .detectors
+            .get(obj_id)
+            .map(|d| (d.track.clone(), d.position)),
+        ObjectType::BufferStop => infra_cache
+            .buffer_stops
+            .get(obj_id)
+            .map(|bs| (bs.track.clone(), bs.position)),
+        _ => None,
+    }
+}
+
+fn matching_endpoints(
```

Comment:
Rename this function:

```suggestion
fn get_matching_endpoint_error(
```